qualification/antenna_channelised_voltage/test_delay.py

```diff
@@ -565,3 +564,163 @@ async def test_phase_rate(
+
+    # Collect about 10s of data, to improve SNR.
+    chunk_timestamp_step = receiver.n_spectra_per_heap * receiver.n_samples_between_spectra
+    n_chunks = round(10.0 * receiver.scale_factor_timestamp // chunk_timestamp_step)
+    n_spectra = n_chunks * receiver.n_spectra_per_heap
+    acc_time = n_chunks * chunk_timestamp_step / receiver.scale_factor_timestamp
+
+    pdf_report.step("Choose a channel.")
+    # Channel is largely arbitrary, although we should avoid the DC frequency
+    channel = receiver.n_chans // 5
+    cfreq = (channel - receiver.n_chans / 2) / receiver.n_chans * receiver.bandwidth + receiver.center_freq
+    pdf_report.detail(f"Using channel {channel}.")
+
+    pdf_report.step("Determine dsim frequency resolution.")
+    dsim_period = await cbf.dsim_clients[0].sensor_value("max-period", int)
+    dsim_resolution = receiver.adc_sample_rate / dsim_period
+    pdf_report.detail(f"Resolution is {dsim_resolution:.6f} Hz.")
+
+    pdf_report.step("Set F-engine gains.")
+    amplitude = 0.8
+    gain = receiver.compute_tone_gain(amplitude, 110)
+    await client.request("gain-all", "antenna-channelised-voltage", gain)
+    pdf_report.detail(f"Set gain on all channels to {gain}.")
+
+    pdf_report.step("Set F-engine phase rate.")
+    # Swing phase through 2pi radians over the collection time. Start away from
+    # phase 0 where quantisation effects are particularly bad.
+    delay_model = f"0,0:0.1,{2 * np.pi / acc_time}"
+    delay_models = [delay_model] * receiver.n_inputs
+    await client.request("delays", "antenna-channelised-voltage", receiver.sync_time, *delay_models)
+    pdf_report.detail(f"Set delays to {delay_model} for all inputs.")
+
+    pdf_report.step("Set beamformer weights to use only one antenna.")
+    weights = np.zeros(len(receiver.source_indices[0]))
+    weights[0] = 1.0
+    await client.request("beam-weights", receiver.stream_names[0], *weights)
+    pdf_report.detail(f"Set weights on {receiver.stream_names[0]}.")
+
+    async def measure_once(freqs: tuple[float, float]) -> tuple[float, float, float]:
+        """Estimate group delay for a single pair of frequencies.
+
+        The delay is ambiguous and could actually be any value of the form
+        :samp:`{delay} + {i} * {period}`.
+
+        Returns
+        -------
+        delay
+            Estimated delay in samples.
+        period
+            Step between possible delay values.
+        std
+            Standard deviation in the delay estimate. Note that the distribution is
+            non-Gaussian and is due to quantisation error, which is bounded.
+        """
+        signal = f"cw({amplitude}, {freqs[0]}); cw({amplitude}, {freqs[1]});"
+        await cbf.dsim_clients[0].request("signals", signal, dsim_period)
+        pdf_report.detail(f"dsim signal set to {signal}.")
+
+        # First axis corresponds to the 2 signals we're comparing.
+        raw_data = np.ones((2, n_spectra, COMPLEX), np.int8)
```

Comment:
I'm curious as to why "ones" and not "zeros"?

Reply:
np.ones ensures that the memory has been written to and hence is paged in (and ideally warm in caches). With large allocations with np.zeros, you might find that the mappings all point at a single read-only zero page and are allocated to "real" pages on the first write, which is not real-time friendly.

However, this is a small allocation where I'm not worried about it. When I started writing the code I was allocating for the entire band rather than one channel. So I'm changing it back to np.zeros rather than adding a comment.